share/server/60/rewrite_fun.js

```diff
@@ -0,0 +1,56 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+//
+// Based on the normalizeFunction which can be
+// found here:
+//
+//  https://github.com/dmunch/couch-chakra/blob/master/js/normalizeFunction.js
+
+function rewriteFunInt(fun) {
+    const ast = esprima.parse(fun);
+    let idx = ast.body.length - 1;
+    let decl = {};
+
+    // Search for the first FunctionDeclaration beginning from the end
+    do {
+        decl = ast.body[idx--];
+    } while (idx >= 0 && decl.type !== "FunctionDeclaration");
+    idx++;
+
+    // If we have a function declaration without an Id, wrap it
+    // in an ExpressionStatement and change it into
+    // a FuntionExpression
+    if (decl.type == "FunctionDeclaration" && decl.id == null) {
```

Comment:
I'm trying to reproduce such a parse tree with the esprima online parser, and I just get errors for things like this:

``` js
function(doc) {}
```

Here's a link: https://esprima.org/demo/parse.html?code=function(doc)%20%7B%7D

Does this actually parse both forms we want it to?

Reply:
It looks that the esprima online parser doesn't like `function(doc) {}`. Using esprima in this PR, we can have output. I tried both `(function (doc) {emit(doc._id, doc._id);});` and `function (doc) {emit(doc._id, doc._id);}`. The result are listed below respectively.

(function (doc) {emit(doc._id, doc._id);});
```
{"type":"Program","body":[{"type":"ExpressionStatement","expression":{"type":"FunctionExpression","id":null,"params":[{"type":"Identifier","name":"doc"}],"body":{"type":"BlockStatement","body":[{"type":"ExpressionStatement","expression":{"type":"CallExpression","callee":{"type":"Identifier","name":"emit"},"arguments":[{"type":"MemberExpression","computed":false,"object":{"type":"Identifier","name":"doc"},"property":{"type":"Identifier","name":"_id"}},{"type":"MemberExpression","computed":false,"object":{"type":"Identifier","name":"doc"},"property":{"type":"Identifier","name":"_id"}}]}}]},"generator":false,"expression":false,"async":false}}],"sourceType":"script"}
```

function (doc) {emit(doc._id, doc._id);}
```
{"type":"Program","body":[{"type":"FunctionDeclaration","id":null,"params":[{"type":"Identifier","name":"doc"}],"body":{"type":"BlockStatement","body":[{"type":"ExpressionStatement","expression":{"type":"CallExpression","callee":{"type":"Identifier","name":"emit"},"arguments":[{"type":"MemberExpression","computed":false,"object":{"type":"Identifier","name":"doc"},"property":{"type":"Identifier","name":"_id"}},{"type":"MemberExpression","computed":false,"object":{"type":"Identifier","name":"doc"},"property":{"type":"Identifier","name":"_id"}}]}}]},"generator":false,"expression":false,"async":false}],"sourceType":"script"}
```

One is `ExpressionStatement` while another is `FunctionDeclaration`

Reply:
@tilgovi The esprima used by this PR is patched to avoid that error so that we can rewrite it. Basically we have to violate the spec as existed in old SpiderMonkey so we can tweak the AST so that it conforms. One of the to-do items we have is to extract that patch so we can apply it to a standard esprima.js to show the diff better.

If memory serves it's something like two lines to remove a boolean check for that "anonymous function at global scope" error.

Reply:
Right, that's the context I was forgetting. Thank you, both!